### PR TITLE
Adding Case Sensitivity to Jaro-Winkler Distance

### PR DESCRIPTION
**Pull request by SJCaldwell**
Other nlp libraries I've seen support caseSensitivity and it seems pretty easy to add, so it might as well be supported. In spellcheck functionality that includes certain names and proper nouns it'll probably be pretty useful. 

The only downside is that caseSensitive is a 4th argument to the JaroWinklerDistance method. Previously you could get away with only passing two parameters and dj would know when it was null, but if you're adding case sensitivity then you have to explicitly declare dj as null, which seems awkward. 

Having caseSensitive as the 3rd argument seemed like it would break existing code though, so I didn't want to change that.

### CODE REVIEW

File: lib/natural/distance/jaro-winkler_distance.js [hunk 1]
```diff
@@ -26,11 +26,14 @@ THE SOFTWARE.
 // http://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance
 // s1 is the first string to compare
 // s2 is the second string to compare
-function distance(s1, s2) {
+// caseSensitive is a boolean expression determining whether 'a' matches 'A'
+function distance(s1, s2, caseSensitive) {
```
Comment (ahmed-masud) [inline lib/natural/distance/jaro-winkler_distance.js]: You should fix this so that any old code still works as expected, i.e. the third flag should be `ignoreCase` so that when it's true you can add new functionality

File: lib/natural/distance/jaro-winkler_distance.js [hunk 2]
```diff
@@ -26,11 +26,14 @@ THE SOFTWARE.
...
 // s1 is the first string to compare
 // s2 is the second string to compare
-function distance(s1, s2) {
+// caseSensitive is a boolean expression determining whether 'a' matches 'A'
+function distance(s1, s2, caseSensitive) {
     if (typeof(s1) != "string" || typeof(s2) != "string") return 0;
     if (s1.length == 0 || s2.length == 0) 
         return 0;
-    s1 = s1.toLowerCase(), s2 = s2.toLowerCase();
+    if (!(caseSensitive))
```
Comment (ahmed-masud) [inline lib/natural/distance/jaro-winkler_distance.js]: Then this would become:

       if(ignoreCase)

**Comment by kkoch986** [comment]
i think this looks great, i do agree with @ahmed-masud changing it so that the old code is not broken would make this much nicer to merge. Also an addition to the README would help raise awareness of this new flag. Let me know if you'd be able to make these changes, i'd still be interested in adding the flag so i might be able to help out at some point if not.

thanks!

**Comment by chrisumbel** [comment]
Regardless of breaking legacy code, is ignoring case considered to be algorithmically incorrect? If so then we *should* break old code. @kkoch986 what thinks you?

**Comment by kkoch986** [comment]
i think i can live with it, let me see what the conflicts look like

**Comment by kkoch986** [comment]
looking at it now, the default as written would not break anything existing. 

the following test passes against the code in this patch:
```javascript
expect(jaroWinklerDistance('aaa', 'aAa', null, true)).not.toBe(jaroWinklerDistance('aaa', 'aaa', null, true));
expect(jaroWinklerDistance('aaa', 'aAa')).toBe(jaroWinklerDistance('aaa', 'aaa'));
```

so if you want case sensitivity you would have to add it. Somehow i had those mixed up in my memory of this issue. It wouldn't be very hard to flip them around though, which it seems like might be the right approach.

we could also do a numeric check on the third argument to give the ability to pass true/false into the third arg and override the case sensitivity. something like this:

```javascript
var ignoreCase = _.isNumber(dj) ? ignoreCase : false;
(typeof(dj) == 'undefined' || !_.isNumber(dj)) ? jaro = distance(s1,s2,caseSensitive) : jaro = dj;
```

not sure how i feel about that but happy to hear what everyone here thinks.

tl;dr:
1. should probably switch caseSensitive to ignoreCase (caseSensitive by default).
1. possibly use the `dj` argument for more than one thing (possibly confusing)

**Comment by MeanwhileMedia** [comment]
Any plans to merge this? I did some rough benchmarking and am seeing approx 17% cpu performance improvement after removing this line:

https://github.com/NaturalNode/natural/blob/master/lib/natural/distance/jaro-winkler_distance.js#L38

**Comment by Hugo-ter-Doest** [comment]
The algorithm as defined on Wikipedia does not say anything about case sensitivity, it is defined in terms of characters. Imho case sensitivity should be the default. Second point about the parameters: I think we should keep things simple, so no double use of parameters. If the user of the method wants to change the default of case sensitivity, then he has to provide a null value for dj as well. (Or he could change the case of the strings before calling the method.)

Hugo

File: lib/natural/distance/jaro-winkler_distance.js (hunk 2 above)
Comment (Hugo-ter-Doest) [reply lib/natural/distance/jaro-winkler_distance.js]: The distance function <code>distance(s1, s2)</code> is not the right spot to change the case of the strings. The strings are also analysed in the main function <code>JaroWinklerDistance()</code>, and changing case here will not be seen by the main function.

It has to be done right at the beginning of the main function:
```javascript
function JaroWinklerDistance(s1, s2, dj, ignoreCase) {
    if (s1 === s2) {
        return 1;
    } else {
        if (ignoreCase) {
          s1 = s1.toLowerCase();
          s2 = s2.toLowerCase();
        }
```

Hugo

**Comment by Hugo-ter-Doest** [comment]
I tried to commit to commit to the fork by Shane, but that didn't work somehow (or I don't know how). Anyway, I will create a new PR with the right corrections, and close this one.

Hugo

**Comment by Hugo-ter-Doest** [comment]
Created and merged [#405](https://github.com/NaturalNode/natural/pull/415) to add this.

Hugo
